### PR TITLE
Set all_reduce_token to None when exiting

### LINKED ISSUE #6246: SIGSEGV when exiting the dataloader in the middle of training
A SIGSEGV error occurs when running the following code with `PJRT_DEVICE=CUDA torchrun --nproc_per_node=2 test.py`.
```
import torch
import torch_xla.core.xla_model as xm
import torch_xla.distributed.parallel_loader as pl

import os
os.environ['PJRT_LOCAL_PROCESS_RANK'] = os.environ['LOCAL_RANK']

device = xm.xla_device()
xm.set_replication(device, [device])
import torch_xla.utils.utils as xu
train_loader = xu.SampleGenerator(
      data=torch.zeros(1, 12),
      sample_count=1024)
train_loader = pl.MpDeviceLoader(train_loader, device)
max_steps = 10
for step, inputs in enumerate(train_loader):
  xm.all_reduce('sum', [inputs], scale=1.0/xm.xrt_world_size())

  if step > max_steps: break
```

This is due to early exit from the dataloader, causing the `xm.mark_step` in pl.MpDeviceLoader to not be executed. As a result, `all_reduce_token` is not set to None, which causes `all_reduce_token` in the global variable `g_all_reduce_tokens` in `torch_xla/csrc/cross_replica_reduces.cpp` to be released only when the program exits.

### PR DESCRIPTION
**Pull request by yitongh**
This pr fixes #6246.

### CODE REVIEW

**Comment by JackCaoG** [comment]
I am curious why does reduce_token matters when we exit the pytorch/xla?

**Comment by JackCaoG** [comment]
oh ok I saw https://github.com/pytorch/xla/issues/6246, my bad. Can you add the test case you mentioned in the issue to a separate test? An example would be https://github.com/pytorch/xla/blob/master/test/test_mp_collective_permute.py which will be run on https://github.com/pytorch/xla/blob/d7c443085f56423bd306eb007882a7cb6f1ac19f/test/run_tests.sh#L231

**Comment by yitongh** [comment]
@JackCaoG I have added a test case, but I cannot run it. The program hangs after setting GPU_NUM_DEVICES=2. Other mp test cases also hang.

**Comment by JackCaoG** [comment]
let's see if CI will be able to run it then.

**Comment by JackCaoG** [comment]
CI seems to be happy, if you can fix the linter I can help you land it.

**Comment by yitongh** [comment]
@JackCaoG I have fixed the linter, sorry forget this.

**Review by JackCaoG — APPROVED** [review]

**Comment by vanbasten23** [comment]
> @JackCaoG I have added a test case, but I cannot run it. The program hangs after setting GPU_NUM_DEVICES=2. Other mp test cases also hang.

This seems to be a new issue. I have also observed the same https://github.com/pytorch/xla/issues/6320. Before this change, all the mp test cases don't hang. @yitongh @JackCaoG

**Comment by JackCaoG** [comment]
@vanbasten23 does this pr breaks multi device GPU training or just the test?

**Comment by JackCaoG** [comment]
@vanbasten23 let's revert his pr. @ManfeiBai can you also help reverting this in the 2.2 release branch? Release date is approaching, I don't want to take this risk.

File: torch_xla/__init__.py [hunk 1]
```diff
@@ -148,6 +148,8 @@ def _setup_tpu_vm_library_path() -> bool:
 
 
 def _prepare_to_exit():
+  device = _XLAC._xla_get_default_device()
```
Comment (alanwaketan) [inline torch_xla/__init__.py:151]: Does this always return the same device regardless of different processes in the pool?

Comment (JackCaoG) [reply torch_xla/__init__.py:151]: I think it will return the device that belong to current process, assuming each process only has one device.

Comment (alanwaketan) [reply torch_xla/__init__.py:151]: Then I'm not sure why it will break other test cases...

Comment (yitongh) [reply torch_xla/__init__.py:151]: I think the error is due to the ComputationClient has exited in atexit._run_exitfuncs when using xla_multiprocessing, while the client still exists when using torchrun. Therefore, in the xla_multiprocessing, this invocation will result in the creation of a new PjRtComputationClient, causing a hang.

 So the better solution is setting the all reduce token in PrepareToExit, like this:
```
diff --git a/torch_xla/__init__.py b/torch_xla/__init__.py
index 8d4997e28..d753f8f7c 100644
--- a/torch_xla/__init__.py
+++ b/torch_xla/__init__.py
@@ -148,8 +148,6 @@ _aws_ec2_inf_trn_init()


 def _prepare_to_exit():
-  device = _XLAC._xla_get_default_device()
-  _XLAC._set_all_reduce_token(device, None)
   _XLAC._prepare_to_exit()
   if int(os.environ.get('PT_XLA_DEBUG', '0')):
     _summarize_fn_tracker()
diff --git a/torch_xla/csrc/init_python_bindings.cpp b/torch_xla/csrc/init_python_bindings.cpp
index 3281f0e9a..b255bb043 100644
--- a/torch_xla/csrc/init_python_bindings.cpp
+++ b/torch_xla/csrc/init_python_bindings.cpp
@@ -97,6 +97,8 @@ void PrepareToExit() {
   runtime::ComputationClient* client =
       runtime::GetComputationClientIfInitialized();
   if (client != nullptr) {
+    auto xla_device = GetDeviceOrCurrent("");
+    SetAllReduceToken(xla_device, nullptr);
     XLAGraphExecutor::Get()->WaitDeviceOps({});
   }
 }
```

**Pull request merged**